src/mlpack/bindings/python/mlpack/matrix_utils.py

```diff
@@ -54,19 +54,26 @@ def to_matrix(x, dtype=np.double, copy=False):
       return x, False
   elif (isinstance(x, np.ndarray) and x.dtype == dtype and x.flags.f_contiguous):
     if copy: # Copy the matrix if required.
-      return np.ndarray(x.shape, buffer=x.flatten(), dtype=dtype, order='C').copy("C"), True
+      return np.ndarray(x.shape, buffer=x.flatten(), dtype=dtype,
+          order='C').copy("C"), True
     else:
-      return np.ndarray(x.shape, buffer=x.flatten(), dtype=dtype, order='C'), False
+      return np.ndarray(x.shape, buffer=x.flatten(), dtype=dtype, order='C'), \
+          False
   else:
     if isinstance(x, pd.core.series.Series) or isinstance(x, pd.DataFrame):
+      # We can only avoid a copy if the dtype is the same and the copy flag is
+      # false.  I'm actually not sure if this is possible, since in everything I
+      # have found, Pandas stores with F_CONTIGUOUS not C_CONTIGUOUS.
       y = x.values
-      if copy: # Copy the matrix if required.
-        return np.ndarray(y.shape, buffer=y.flatten(), dtype=dtype, order='C').copy("C"), True
+      if copy == False and y.dtype == dtype and y.flags.c_contiguous:
+        return np.ndarray(y.shape, buffer=y.flatten(), dtype=dtype, order='C'),\
```

Comment:
Hey @rcurtin if there is any problem if we pass Pandas matrix (`F_CONTIGUOUS`) here. Because I think it will convert `F_CONTIGUOUS` into `C_CONTIGUOUS`without creating any copy, which we can pass correctly to mlpack.

```
>>> x = pd.read_csv("iris.csv")
>>> y = x.values
>>> y.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False
>>> z = np.ndarray(y.shape, buffer=y.flatten(), dtype=y.dtype, order='C')
>>> z.flags
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False
>>> (y == z).all()
True
```

Reply:
I spent a while confused and staring at the Python terminal here before I realized what happened.  When you do this, take a look and see that `OWNDATA` for *both* `y` and `z` are `False`, which is surprising!  `y` is definitely just using the memory of `x`, so that is not surprising.  But what memory is `z` using?  It cannot be the same memory as `y` because the order is different.

Then I realized that when we call `y.flatten()`, there is a default parameter: `order='C'`.  If the matrix given to `flatten()` is not in the desired order, a copy with the new order will be made.  So actually `z` is using the memory of the result of `y.flatten()`, which is now `C_CONTIGUOUS`!

So I think the code above is fine, since `flatten()` will not copy when `order='C'` (which it is in the code above).  The `else` below just copies directly, which is the same end result.

Let me know if I missed anything... it was a bit difficult to dig in and figure out what was going on.

Reply:
Thanks @rcurtin for clarification. :) The above is looking good to me as well.